### DISCUSSION TITLE
Highspy, logging options?

### DISCUSSION THREAD

**Discussion by RobinBergmark** [ERGO-Code/HiGHS#1833] [General]
Hi,

I've been using Highspy in an app. Works great in general, but I'm wondering if there are any additional options for logging? Specifically, I log everything from the web sever (python code) as json. But it doesn't look to me as if that would be possible with the type of output HiGHS logs? Other things I'd be interested in is to add contextual information to each log entry (e.g. request id and such), is that possible?

Per the documentation it looks like you can just set whether to log to console/file unless I missed something, so I'm wondering if there are any other clever ideas? Otherwise I'll just log it to a file and take it from there, but it would be convenient if there was something else. 

Thanks for any insights!

**Comment by jajhall** [comment]
Your only option for this level of control and contextualisation is to use the logging callback. See

https://ergo-code.github.io/HiGHS/dev/callbacks/

**Reply by RobinBergmark** [reply]
Hi, and thanks for the reply! I looked at the callbacks and it works mostly fine. Most of the lines that get logged now pass through the callback and I can format them however I want. However, some specific parts of the logging seem unaffected. Specifically the summary that it logs at the end: 


```
| Coefficient ranges:
|   Matrix [1e-05, 1e+05]
|   Cost   [3e-01, 2e+04]
|   Bound  [1e+00, 5e+05]
|   RHS    [1e+00, 7e+04]
| WARNING: No semi-integer/integer variables in model with non-empty integrality
| Presolving model
| 1541 rows, 1899 cols, 4470 nonzeros  1s
| 781 rows, 1236 cols, 3265 nonzeros  1s
| 668 rows, 1201 cols, 2947 nonzeros  1s
| Presolve : Reductions: rows 668(-6930); columns 1201(-5222); elements 2947(-17642)
| Solving the presolved LP
| Using EKK dual simplex solver - serial
|   Iteration        Objective     Infeasibilities num(sum)
|           0     2.1827315716e+01 Pr: 425(1.54287e+06) 2s
|         630     1.2507371516e+06 Pr: 0(0) 2s
| Solving the original LP from the solution after postsolve
| Model   status      : Optimal
| Simplex   iterations: 630
| Objective value     :  1.2507471516e+06
| HiGHS run time      :          1.80
```

Aside from that, it's the initial log line of "Running HiGHS 1.7.1 (git hash: 0c240d8): Copyright (c) 2024 HiGHS under MIT licence terms" which also gets printed as plain text, and does not seem to get pass through the callback .

Is this by intent, or something that needs further configuration? Or is it a bug that I should raise an issue for?

**Reply by jajhall** [reply]
There are a few things that don't appear through the I/O mechanism that calls the logging callback, and fixing this is on my list of issues.

**Reply by RobinBergmark** [reply]
Okay, that's nice to know. Thanks for the help, and the fast replies!
